### ISSUE TITLE
No Qt6 platform plugin could be initialized

### ISSUE THREAD

**Issue by embeddedmz** [microsoft/vcpkg#20101]
**Describe the bug**
When launching a Qt6 program on VS2019 (Solution has been generated via CMake), I got a Debug Error with this message : 
```
qt.qpa.plugin: Could not find the Qt platform plugin "windows" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

Screenshot of the error dialog box : https://ibb.co/DWF4swM

There's no issue when launching Qt5 programs. I need to use Qt6 because the Qt5 version provided by vcpkg has this bug : https://forum.qt.io/topic/130087/adjust-qtableview-height-to-its-content-few-lines that doesn't exist with the latest Qt5 libraries.

**Environment**
- OS: Windows 10
- Compiler: VS2019

**Comment by Neumann-A** [comment]
You either need to run `windeployqt` or have a `qt.conf` alongside your application with the correct paths setup.

**Comment by embeddedmz** [comment]
I found a windeployqt.exe inside :
D:\vcpkg\installed\x64-windows\tools\Qt6\bin
D:\vcpkg\packages\qttools_x64-windows\tools\Qt6\bin
D:\vcpkg\buildtrees\qttools\x64-windows-rel\bin

There's this Qt documentation https://doc.qt.io/qt-5/windows-deployment.html but this is ain't simple. I'm gonna stick to Qt5 :( I want Qt6 to work out of the box ! Do you think uninstalling Qt5 can help ?

**Comment by Neumann-A** [comment]
@FranciscoPombal: Can you link some code which invokes windeployqt?

**Comment by stijnherfst** [comment]
So is this something that is now being treated as a bug seeing as it is tagged category:port-bug? It is painful to use Qt6 now as you have to either copy a bunch of files (manually) or run windeployqt (slow).

**Comment by JackBoosY** [comment]
@Neumann-A Do we have any ways to automaticly call those scripts?

@stijnherfst  Since the scripts are provided by qt ports, I marked port bug temporary. Also I can mark this as tool update.

**Comment by Neumann-A** [comment]
> Do we have any ways to automaticly call those scripts?

vcpkg could but the question is rather if vcpkg should. The build/deploy scripts should be independent of vcpkg and this change would make the downstream deploy scripts incomplete. I consider this a learning opportunity for the user on how to properly setup qt projects..... (also you can complain upstream that https://doc.qt.io/qt-6/cmake-manual.html does not contain infos on deployment. )

>  It is painful to use Qt6 now as you have to either copy a bunch of files (manually) or run windeployqt (slow).

simply use static library linkage for development if you don't want to deal with proper deployment? As long as you only internally use the produced binaries you should be fine. (Also windeployqt is kind of required even if you think it is slow)

**Comment by stijnherfst** [comment]
Thank you for your reply.

>  (also you can complain upstream that https://doc.qt.io/qt-6/cmake-manual.html does not contain infos on deployment. )
> simply use static library linkage for development if you don't want to deal with proper deployment?
I wasn't specifically talking about deployment, but rather development. Running windeployqt is no issue once the application is ready to be deployed, but in the development phase it would be terribly slow/annoying to have to run windeployqt on every compilation. 

The plugins are also found under `vcpkg\installed\x64-windows\Qt6\plugins`. Were they put there for a reason such as having them in the location vcpkg uses to copy over binaries?

This plugin thing was not an issue with vcpkg + Qt5 so you could even say this is a regression. There it would automatically take care of the plugin DLLs.

**Comment by Neumann-A** [comment]
> This plugin thing was not an issue with vcpkg + Qt5 so you could even say this is a regression. There it would automatically take care of the plugin DLLs.

There is a powershell script in Qt5 which tries to mimic windeployqt. But a) this is reinventing the wheel since qtwindeployqt exists and b) needs additional maintenance, that script is not perfectly copying windeployqt and has known bugs (especially in regards to qml).  
Also this is not a regression since qt6 is a different port than qt5. 

> it would be terribly slow/annoying to have to run windeployqt on every compilation

It would be? Have you tried? the qt5 powershell skript is trying to do the same thing as windeployqt and that is running dumpbin on the executable to get a list of the dependent dlls (recursively). If you don't want it to run on every compilation make it a custom target instead of a post build event.

**Comment by stijnherfst** [comment]
> It would be? Have you tried? 
I'll shamefully admit that I haven't for this project and was making the assumption based on previous experience with windeployqt. Would the recommended solution to this issue then be to have users call windeployqt (or Mac equivalent) post build via CMake?

**Comment by JackBoosY** [comment]
Should we print a message to the user?

**Comment by JackBoosY** [comment]
Does this issue resolved?

**Comment by embeddedmz** [comment]
I don't think so, I will check it next week.

**Comment by embeddedmz** [comment]
I failed in installing qt6 :

```
PS D:\vcpkg> .\vcpkg install qt
Computing installation plan...
Package qtlottie does not have a default-features feature
Note: Updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.
```

**Comment by JackBoosY** [comment]
@embeddedmz Sorry for late, please remove `qtlottie` and try again.

**Comment by embeddedmz** [comment]
@JackBoosY I'm still having the Debug error message box.

I tried using this project : https://github.com/embeddedmz/QTableViewAdjustPolicyNotWorkingProperly

By the way please change this line : 
find_package(QT NAMES Qt5 Qt6 COMPONENTS Widgets REQUIRED)
into
find_package(QT NAMES Qt6 COMPONENTS Widgets REQUIRED)

So you can use Qt6 instead of Qt5.

**Comment by embeddedmz** [comment]
@JackBoosY  vcpkg failed to complete building Qt because of qtwebengine, I restarted installing Qt and I got this :

```
PS D:\vcpkg> .\vcpkg.exe install qt:x64-windows
Computing installation plan...
The following packages will be built and installed:
    qt[core,default-features]:x64-windows -> 6.2.3
  * qtwebengine[core,default-features,geolocation,spellchecker,webchannel]:x64-windows -> 6.2.3
  * qtwebsockets[core,qml]:x64-windows -> 6.2.3
  * qtwebview[core,qml,webengine]:x64-windows -> 6.2.3
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
Restored 0 packages from C:\Users\mmzoughi\AppData\Local\vcpkg\archives in 1.411 ms. Use --debug to see more details.
Starting package 1/4: qtwebengine:x64-windows
Building package qtwebengine[core,default-features,geolocation,spellchecker,webchannel]:x64-windows...
-- Using cached D:/vcpkg/downloads/qtwebengine-30c22c6ed9833c7e6e14a345752c6f13cfbaec51.tar.gz
-- Using source at D:/vcpkg/buildtrees/qtwebengine/src/13cfbaec51-a3418e0ed7.clean
-- Using cached D:/vcpkg/downloads/qtwebengine-855304132f321f285986c7f1710a45bae72aec12.tar.gz
-- Cleaning sources at D:/vcpkg/buildtrees/qtwebengine/src/bae72aec12-bb8d63e444.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/vcpkg/downloads/qtwebengine-855304132f321f285986c7f1710a45bae72aec12.tar.gz
-- Using source at D:/vcpkg/buildtrees/qtwebengine/src/bae72aec12-bb8d63e444.clean
-- Found external ninja('1.10.2').
-- Configuring x64-windows-dbg
-- Configuring x64-windows-rel
-- Building x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:158 (message):

Error: Building package qtwebengine:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `git pull` and `.\vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+qtwebengine
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[qtwebengine]+Build+error
including:
  package: qtwebengine[core,default-features,geolocation,spellchecker,webchannel]:x64-windows -> 6.2.3
    vcpkg-tool version: 2022-03-09-1affd32f93b299d5a907816c328ca3ededb73a7e
    vcpkg-scripts version: 965961552 2022-03-14 (2 days ago)
```

**Comment by Neumann-A** [comment]
@embeddedmz if you don't need qtwebengine don't install it. Only install components you actually use. otherwise you need ninja with long path support or add `--x-buildtree-root=D:\b\` to  the install command

**Comment by embeddedmz** [comment]
@JackBoosY @Neumann-A I still have the same error :
![vcpkgQt6Error](https://user-images.githubusercontent.com/10297393/158563608-d5073485-2033-449c-a2ea-2400d1221b1c.png)

**Comment by Neumann-A** [comment]
@embeddedmz: You need to deploy the Qt plugins (either manually or using windeployqt) or have a `qt.conf` with the correct paths setup. If you don't know how to do that I would advise you to stick to `-static` triplets for Qt on windows.

**Comment by embeddedmz** [comment]
@Neumann-A why don't I have this problem with Qt 5 ?

To fix it manually, what do I have to copy and where ?

Thank you

**Comment by Neumann-A** [comment]
https://stackoverflow.com/questions/20495620/qt-5-1-1-application-failed-to-start-because-platform-plugin-windows-is-missi

**Comment by embeddedmz** [comment]
@Neumann-A  I did a search for "platforms" in the vcpkg folder and tried to copy the DLLs on the build's output directory but it didn't work (I tried different locations...).

Calling windeployqt (found on the vcpkg folders) on the Release output build directory made the executable working.

D:\vcpkg\buildtrees\qttools\x64-windows-dbg\bin\windeployqt.exe does not display anything on the screen and do not deploy Qt for the Debug version of the program 👎

**Comment by Neumann-A** [comment]
> Calling windeployqt (found on the vcpkg folders) on the Release output build directory made the executable work.

See it works if you did it correctly. And now do it from CMake using the windeploy target and it will magically work for Release **AND** Debug.

> Debug folder as a first argument and do not deploy Qt for the Debug version of the program 👎

Yeah because you clearly don't understand how windeploy works. There is a `windeployqt.debug.bat` which is ment for debug builds..... and in a file explorer it is one item above windeployqt.exe. (Short story windeployqt searches for qmake and qmake can only use one qt.conf at a time which is why vcpkg has to use a batch file to tell windeployqt to use a different qmake which uses a different qt.conf; would be much easier if windeployqt would just accept a qt.conf as input. )

> Qt6 is still not well implemented in vcpkg unlike Qt5

This is not a vcpkg issue. It is a Qt issue. vcpkg did hold your hand with Qt5 a bit too much but Qt6 won't do that for you. Consider it a learning experience how to correctly deploy your binaries.
However be aware that starting with Qt 6.3 you might get https://doc-snapshots.qt.io/qt6-dev/qt-deploy-runtime-dependencies.html which helps with the issue at hand.

**Comment by genevanmeter** [comment]
In the thick of this now the best I can offer is 
vcpkg install qttools or add to manifest and:
```
find_program(TOOL_WINDEPLOYQT NAMES windeployqt)

add_custom_command(TARGET myApp POST_BUILD
    COMMAND ${TOOL_WINDEPLOYQT}
            $<TARGET_FILE:myApp>
    COMMENT "Running windeployqt..."
)
```

Which only seems to work in Release.  In Debug it can't find the dlls (Qt6***d.dll) because it is looking in bin/ rather than debug/bin/ but I have been unable to override it.

**Comment by lukase** [comment]
> Yeah because you clearly don't understand how windeploy works. There is a `windeployqt.debug.bat` which is ment for debug builds..... and in a file explorer it is one item above windeployqt.exe.

This batch file is not being deployed for me (vcpkg on commit 94c728df6, in a project using a manifest listing 'qtbase' and 'qttools' as dependency). Looks like windeployqt has been removed from the qttools package into qtbase. Qt version being built is 6.3.0

**Comment by JackBoosY** [comment]
Is this issue resolved?

**Comment by JackBoosY** [comment]
We hope your question was answered to your satisfaction; if it wasn't, you can reopen with more info.

**Comment by kwsp** [comment]
This issue hasn't been resolved, can we please reopen?

**Comment by endingly** [comment]
> 
Excuse me, after running the command here, windeployqt.exe told me that the subsequent application is not a QT program because of what reason? In fact, the following application is an opencv program with QT support enabled, which should logically be a QT program.
![image](https://user-images.githubusercontent.com/38353844/231924342-1e05ac00-9d02-4050-b354-64b87dcae9bb.png)

**Comment by QiuYilin** [comment]
> In the thick of this now the best I can offer is vcpkg install qttools or add to manifest and:
> 
> ```
> if (CMAKE_BUILD_TYPE STREQUAL "Debug")
> 	find_program(TOOL_WINDEPLOYQT NAMES windeployqt.debug.bat)
> else()
> 	find_program(TOOL_WINDEPLOYQT NAMES windeployqt)
> endif()
> 
> add_custom_command(TARGET myApp POST_BUILD
>     COMMAND ${TOOL_WINDEPLOYQT}
>             $<TARGET_FILE:myApp>
>     COMMENT "Running windeployqt..."
> )
> ```
> 
> ~Which only seems to work in Release. In Debug it can't find the dlls (Qt6***d.dll) because it is looking in bin/ rather than debug/bin/ but I have been unable to override it.~
> 
> Updated to work with Debug.

CMAKE_BUILD_TYPE  not work in MSVC

**Comment by Neumann-A** [comment]
> CMAKE_BUILD_TYPE not work in MSVC

Then use the `windeployqt` imported target. (Don't have a installation at hand currently so cannot check the correct name, however I know that there is one imported target for windeployqt. )

**Comment by kelteseth** [comment]
My fix was simply to tell Qt where to find all imports and stuff via `qt.conf`. It only worked after adding it next to my executable. You may have to alter your **Prefix** paths:

```
[DevicePaths]
Prefix=./../../vcpkg/installed/x64-windows
Headers=include/Qt6/
Libraries=lib
Plugins=Qt6/plugins
Qml2Imports=Qt6/qml
Documentation=doc/Qt6/
Binaries=bin
LibraryExecutables=tools/Qt6/bin
ArchData=share/Qt6
Data=share/Qt6
Translations=translations/Qt6/
Examples=share/examples/Qt6/
[Paths]
Prefix=./../../vcpkg/installed/x64-windows
Headers=include/Qt6/
Libraries=lib
Plugins=Qt6/plugins
Qml2Imports=Qt6/qml
Documentation=doc/Qt6/
Binaries=bin
LibraryExecutables=tools/Qt6/bin
ArchData=share/Qt6
Data=share/Qt6
Translations=translations/Qt6/
Examples=share/examples/Qt6/
HostPrefix=./../../vcpkg/installed/x64-windows
HostData=share/Qt6
HostBinaries=bin
HostLibraries=lib
HostLibraryExecutables=tools/Qt6/bin

```

**Comment by kelteseth** [comment]
@Neumann-A so my current CMake projects looks like this:
```
# ######### Copy files ##########
# Since we vcpkg qt we need to tell Qt where to find qml plugins and files
# and other paths like where to find the QtWebEngineProcess executable.
# This is only needed for dev builds
if(CMAKE_BUILD_TYPE STREQUAL "Debug")
    configure_file(qt.debug.conf ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/qt.conf COPYONLY)
else()
    configure_file(qt.conf ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/qt.conf COPYONLY)
endif()
```
I do need the two different config file, so I do not get a runtime error stating that it cannot locate QtWebEngineProcess.exe

This has now several drawbacks

1. This is hardcoded for windows and x64
2. I have to invoke it myself
3. I have to remove it again when deploying

Theres got to be a better way for doing this, right? I tried to add the paths where the qt.conf is located to my vscode launch.json env variable, but no combination worked for me. Only copying next to me executable worked.

**Comment by acdemiralp** [comment]
It is absolutely insane that no clean solution exists to this problem in 2026. To clear things for you as a first step, this IS a problem.
